skimage/feature/_canny.py

```diff
@@ -14,44 +14,176 @@
 
 import numpy as np
 import scipy.ndimage as ndi
-from scipy.ndimage import generate_binary_structure, binary_erosion, label
+
 from ..filters import gaussian
-from .. import dtype_limits, img_as_float
+from .. import dtype_limits
 from .._shared.utils import check_nD
 
 
-def smooth_with_function_and_mask(image, function, mask):
-    """Smooth an image with a linear function, ignoring masked pixels.
+def _preprocess(image, mask, sigma, mode):
+    """Generate a smoothed image and an eroded mask.
+
+    The image is smoothed using a gaussian filter ignoring masked
+    pixels and the mask is eroded.
 
     Parameters
     ----------
     image : array
-        Image you want to smooth.
-    function : callable
-        A function that does image smoothing.
+        Image to be smoothed.
     mask : array
         Mask with 1's for significant pixels, 0's for masked pixels.
+    sigma : scalar or sequence of scalars
+        Standard deviation for Gaussian kernel. The standard
+        deviations of the Gaussian filter are given for each axis as a
+        sequence, or as a single number, in which case it is equal for
+        all axes.
+    mode : str, {'reflect', 'constant', 'nearest', 'mirror', 'wrap'}
+        The ``mode`` parameter determines how the array borders are
+        handled, where ``cval`` is the value when mode is equal to
+        'constant'. Default is 'nearest'.
+
+    Returns
+    -------
+    smoothed_image : ndarray
+        The smoothed array
+    eroded_mask : ndarray
+        The eroded mask.
 
-    Notes
-    ------
-    This function calculates the fractional contribution of masked pixels
-    by applying the function to the mask (which gets you the fraction of
-    the pixel data that's due to significant points). We then mask the image
-    and apply the function. The resulting values will be lower by the
-    bleed-over fraction, so you can recalibrate by dividing by the function
-    on the mask to recover the effect of smoothing from just the significant
-    pixels.
     """
-    bleed_over = function(mask.astype(float))
-    masked_image = np.zeros(image.shape, image.dtype)
+
+    if mask is None:
+        mask = np.ones(image.shape, dtype=bool)
```

Comment:
I think I prefer to see these two lines in the main function (that defines a default behavior)

Reply:
We don't really need `mask` in the main function! Defining it there will keep it in memory until the end of the function call :confused: